api/controllers/course.go

```diff
@@ -0,0 +1,81 @@
+package controllers
+
+import (
+	"context"
+	"net/http"
+	"time"
+
+	"github.com/UTDNebula/nebula-api/api/configs"
+	"github.com/UTDNebula/nebula-api/api/responses"
+
+	"github.com/gin-gonic/gin"
+
+	"go.mongodb.org/mongo-driver/bson"
+	"go.mongodb.org/mongo-driver/bson/primitive"
+	"go.mongodb.org/mongo-driver/mongo"
+)
+
+var courseCollection *mongo.Collection = configs.GetCollection(configs.DB, "courses")
+
+func CourseSearch() gin.HandlerFunc {
+	return func(c *gin.Context) {
+		//name := c.Query("name")            // value of specific query parameter: string
+		queryParams := c.Request.URL.Query() // map of all query params: map[string][]string
+
+		ctx, cancel := context.WithTimeout(context.Background(), 10*time.Second)
+
+		// @TODO: Fix with model - There is NO typechecking!
+		// var courses []models.Course
+		var courses []map[string]interface{}
+
+		defer cancel()
+
+		// build query key value pairs (only one value per key)
+		query := bson.M{}
+		for key, _ := range queryParams {
+			query[key] = c.Query(key)
+		}
+
+		// get cursor for query results
+		cursor, err := courseCollection.Find(ctx, query)
+		if err != nil {
+			c.JSON(http.StatusInternalServerError, responses.CourseResponse{Status: http.StatusInternalServerError, Message: "error", Data: err.Error()})
```

Comment:
My understanding is that the best practice is to implement application level query/document validation. In our case, we are building the (relatively basic) queries for matching the provided key/value pairs. Therefore to the user this is always an internal server error.

Reply:
Ideally you should have query validation on both ends. Regardless, you should return a BadRequest if there's a problem with the query so that client-side developers understand that it's most likely an issue on their end and not an issue on ours.

Reply:
Ideally you should have query validation on both ends. Regardless, you should return a BadRequest if there's a problem with the query so that client-side developers understand that it's most likely an issue on their end and not an issue on ours.

Reply:
My understanding (based on what you've said and general assumptions. I've been unable to find a definitive source on the matter. Please forward one if you can find out) is that an error here would occur if an issue with the servers' ability to query (i.e. database goes down), or there is an issue with the query (formatting, should never run into this using built-ins). Since we are building the actual query, the latter would be an issue on our end, as would an issue if the database went down. If the user input incorrect query data, they would receive the incorrect (often empty) data successfully.

Reply:
Got it that makes sense. I'll approve it - I'll probably just have to educate myself on mongo a little more.